src/gradualizer_lib.erl

```diff
@@ -83,58 +84,105 @@ reverse_graph(G) ->
     from_edges(maps:keys(G), [ {J, I} || {I, Js} <- maps:to_list(G), J <- Js ]).
 
 
-% Given a type, pick a value of that type.
-% Used in exhaustiveness checking to show an example value
-% which is not covered by the cases.
-pick_value(List) when is_list(List) ->
-    [pick_value(Ty) || Ty <- List ];
-pick_value(?type(integer)) ->
+%% Look up `UserTy' definition:
+%% first in `gradualizer_db', then, if not found, in provided `Types' map.
+%% `UserTy' is actually an unexported `gradualizer_type:af_user_defined_type()'.
+
+-spec get_user_type_definition(UserTy, Types) -> {ok, Ty} | opaque | not_found when
+      UserTy :: gradualizer_type:abstract_type(),
+      Types :: #{{Name :: atom(), arity()} => {Params :: [atom()],
+                                               Body :: gradualizer_type:abstract_type()}},
+      Ty :: gradualizer_type:abstract_type().
+get_user_type_definition({user_type, Anno, Name, Args}, Types) ->
+    %% Let's check if the type is a known remote type.
+    case typelib:get_module_from_annotation(Anno) of
+        {ok, Module} ->
+            case gradualizer_db:get_type(Module, Name, Args) of
+                {ok, Ty} ->
+                    {ok, Ty};
+                opaque ->
+                    opaque;
+                not_found ->
+                    not_found
+            end;
+        none ->
+            %% Let's check if the type is defined in the context of this module.
+            case maps:get({Name, length(Args)}, Types, not_found) of
+                {_Params, Ty} ->
+                    {ok, Ty};
+                not_found ->
+                    not_found
+            end
+    end.
+
+
+%% Given a type `Ty', pick a value of that type.
+%% Used in exhaustiveness checking to show an example value
+%% which is not covered by the cases.
+
+-spec pick_value(Ty, Types) -> AbstractVal when
+      Ty :: gradualizer_type:abstract_type(),
+      Types :: #{{Name :: atom(), arity()} => {Params :: [atom()],
+                                               Body :: gradualizer_type:abstract_type()}},
+      AbstractVal :: gradualizer_type:abstract_expr().
```

Comment:
Types might be called UserTypes, UserTypeMap or something (only in the spec) to avoid confusion Ty/Types.